samples/net/coap_client/src/main.c

```diff
@@ -0,0 +1,395 @@
+
+	err = coap_packet_init(&request, coap_buf, sizeof(coap_buf),
+			       APP_COAP_VERSION, COAP_TYPE_NON_CON,
+			       sizeof(next_token), (uint8_t *)&next_token,
+			       COAP_METHOD_GET, coap_next_id());
+	if (err < 0) {
+		LOG_ERR("Failed to create CoAP request, %d", err);
+		return err;
+	}
+
+	err = coap_packet_append_option(&request, COAP_OPTION_URI_PATH,
+					(uint8_t *)CONFIG_COAP_RESOURCE,
+					strlen(CONFIG_COAP_RESOURCE));
+	if (err < 0) {
+		LOG_ERR("Failed to encode CoAP option, %d", err);
+		return err;
+	}
+
+	err = send(sock, request.data, request.offset, 0);
+	if (err < 0) {
+		LOG_ERR("Failed to send CoAP request, %d", errno);
+		return -errno;
+	}
+
+	LOG_INF("CoAP request sent: token 0x%04x", next_token);
+
+	return 0;
+}
+
+/**
+ * @brief Waits until data is available on the socket
+ *
+ * @return 0 on success
+ * @return -EAGAIN if timeout occurred and there is no data.
+ * @return negative error code in case of poll error.
+ */
+static int wait(int timeout)
+{
+	int ret = poll(&fds, 1, timeout);
+
+	if (ret < 0) {
+		LOG_ERR("poll error: %d", errno);
+		return -errno;
+	}
+
+	if (ret == 0) {
+		/* Timeout. */
+		return -EAGAIN;
+	}
+
+	if ((fds.revents & POLLERR) == POLLERR) {
+		LOG_ERR("wait: POLLERR");
+		return -EIO;
+	}
+
+	if ((fds.revents & POLLNVAL) == POLLNVAL) {
+		LOG_ERR("wait: POLLNVAL");
+		return -EBADF;
+	}
+
+	if ((fds.revents & POLLIN) != POLLIN) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Waits for the establishment of a network connection.
+ */
+static void wait_for_network(void)
+{
+	k_mutex_lock(&network_connected_lock, K_FOREVER);
+	if (!is_connected) {
+		LOG_INF("Waiting for network connectivity");
+		k_condvar_wait(&network_connected, &network_connected_lock, K_FOREVER);
+	}
+	k_mutex_unlock(&network_connected_lock);
+}
+
+/**
+ * @brief Infinite CoAP request loop
+ */
+static void coap_loop(void)
+{
+	int64_t next_msg_time = APP_COAP_SEND_INTERVAL_MS;
+	int err, received;
+
+	wait_for_network();
+
+	if (server_resolve() != 0) {
+		LOG_ERR("Failed to resolve server name");
+		return;
+	}
+
+	if (client_init() != 0) {
+		LOG_ERR("Failed to initialize CoAP client");
+		return;
+	}
+
+	next_msg_time = k_uptime_get();
+
+	while (1) {
+		wait_for_network();
+
+		if (k_uptime_get() >= next_msg_time) {
+			if (client_get_send() != 0) {
+				LOG_ERR("Failed to send GET request, exit...");
+				break;
+			}
+
+			next_msg_time = k_uptime_get() + APP_COAP_SEND_INTERVAL_MS;
+		}
+
+		int64_t remaining = next_msg_time - k_uptime_get();
+
+		if (remaining < 0) {
+			remaining = 0;
+		}
+
+		err = wait(remaining);
+		if (err < 0) {
+			if (err == -EAGAIN) {
+				continue;
+			}
+
+			LOG_ERR("Poll error...");
+			continue;
+		}
+
+		received = recv(sock, coap_buf, sizeof(coap_buf), MSG_DONTWAIT);
+		if (received < 0) {
+			if (errno == EAGAIN || errno == EWOULDBLOCK) {
+				LOG_ERR("socket EAGAIN");
+				continue;
+			} else {
+				LOG_ERR("Socket error...");
+				continue;
+			}
```

Comment:
I don't think we should handle `EAGAIN` differently than any other errors at this point, given that we checked the socket had POLLIN beforehand. I would remove one branch and print the `errno`.

Reply:
Also fixed!